File: JavaScript/ExampleTests/common/MatchTyped.js
import { MatchString } from "./MatchString.js";
import * as _json from "./json.js";
import { AddressComponentsTyped } from "./AddressComponentsTyped.js";

export class MatchTyped {
  constructor(matchedAddress = "", addressComponents = "") {
    this.matchedAddress = matchedAddress;
    this.addressComponents = addressComponents;
  }

  static fromStringObj(s) {
    return new MatchTyped(
      s.matchedAddress,
      AddressComponentsTyped.fromStringObj(s.addressComponents)
    );
  }

  toStringObj() {
    return new MatchString(
      String(this.matchedAddress),
      this.addressComponents.toStringObj()
    );
  }

  static toStringList(list) {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list) {
    return list.map(s => MatchTyped.fromStringObj(s));
  }

  toJsonValue() {
    return {
      matchedAddress: this.matchedAddress,
      addressComponents: this.addressComponents.toJsonValue(),
    };
  }

  toJSON() { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m) {
    return new MatchTyped(
      _json.asString(_json.require(m, "matchedAddress"), "matchedAddress"),
      AddressComponentsTyped.fromJsonValue(_json.require(m, "addressComponents"))
    );
  }

  static fromJSON(text) { return MatchTyped.fromJsonValue(_json.parse(text)); }

  static toJSONList(items) {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text) {
    const raw = _json.asArray(_json.parse(text), "MatchTyped");
    return raw.map((e) => MatchTyped.fromJsonValue(e));
  }

  toString() {
    return `matchedAddress=${this.matchedAddress}, addressComponents=${this.addressComponents}`;
  }

  equals(other) {
    if (!(other instanceof MatchTyped)) return false;
    return this.matchedAddress === other.matchedAddress
      && this.addressComponents.equals(other.addressComponents);
  }
}

File: Rust/ExampleTests/common/match__typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::match__string::MatchString;
use super::addresscomponents_typed::AddressComponentsTyped;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct MatchTyped {
    pub matchedaddress: String,
    pub addresscomponents: AddressComponentsTyped,
}

impl MatchTyped {
    pub fn from_str_struct(s: &MatchString) -> Self {
        Self {
            matchedaddress: s.matchedaddress.clone(),
            addresscomponents: AddressComponentsTyped::from_str_struct(&s.addresscomponents),
        }
    }

    pub fn to_str_struct(&self) -> MatchString {
        MatchString {
            matchedaddress: self.matchedaddress.clone(),
            addresscomponents: self.addresscomponents.to_str_struct(),
        }
    }

    pub fn to_string_list(list: &[MatchTyped]) -> Vec<MatchString> {
        list.iter().map(|t| t.to_str_struct()).collect()
    }

    pub fn from_string_list(list: &[MatchString]) -> Vec<MatchTyped> {
        list.iter().map(MatchTyped::from_str_struct).collect()
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("matchedAddress".to_string(), json::Value::Str(self.matchedaddress.clone())),
            ("addressComponents".to_string(), self.addresscomponents.to_json_value()),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            matchedaddress: json::as_string(json::require(v, "matchedAddress")?, "matchedAddress")?,
            addresscomponents: AddressComponentsTyped::from_json_value(json::require(v, "addressComponents")?)?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[MatchTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<MatchTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "MatchTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
